include/deal.II/lac/block_linear_operator.h

```diff
@@ -26,23 +26,365 @@
+      Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+      Assert(u.n_blocks() == n, ExcDimensionMismatch(u.n_blocks(), n));
+
+      for (size_t i = 0; i < m; ++i)
+        for (size_t j = 0; j < n; ++j)
+          op.block(i, j).vmult_add(v.block(i), u.block(j));
+    };
+
+    op.Tvmult = [=](Domain &v, const Range &u)
+    {
+      const unsigned int n = op.n_block_cols();
+      const unsigned int m = op.n_block_rows();
+      Assert(v.n_blocks() == n, ExcDimensionMismatch(v.n_blocks(), n));
+      Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+      for (size_t i = 0; i < n; ++i)
+        {
+          op.block(0, i).Tvmult(v.block(i), u.block(0));
+          for (size_t j = 1; j < m; ++j)
+            op.block(j, i).Tvmult_add(v.block(i), u.block(j));
+        }
+    };
+
+    op.Tvmult_add = [=](Domain &v, const Range &u)
+    {
+      const unsigned int n = op.n_block_cols();
+      const unsigned int m = op.n_block_rows();
+      Assert(v.n_blocks() == n, ExcDimensionMismatch(v.n_blocks(), n));
+      Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+      for (size_t i = 0; i < n; ++i)
+        for (size_t j = 0; j < m; ++j)
+          op.block(j, i).Tvmult_add(v.block(i), u.block(j));
+    };
+  }
+
+} /*namespace*/
+
+
+
+/**
+ * @name Creation of a BlockLinearOperator
  */
 //@{
 
 /**
  * @relates LinearOperator
  *
- * A function that encapsulates a given collection @p ops of LinearOperators
- * into a block structure. Hereby, it is assumed that Range and Domain are
- * blockvectors, i.e., derived from
- * @ref BlockVectorBase.
- * The individual linear operators in @p ops must act on a the underlying
- * vector type of the block vectors, i.e., on Domain::BlockType yielding a
- * result in Range::BlockType.
+ * A function that encapsulates a @p block_matrix into a
+ * BlockLinearOperator.
+ *
+ * All changes made on individual blocks of @p block_matrix after the
+ * creation of the BlockLinearOperator object are reflected by the operator
+ * object. The block structure of @p block_matrix, however, must not be
+ * changed.
+ *
+ * @ingroup LAOperators
+ */
+template <typename Range,
+          typename Domain,
+          typename BlockMatrix>
+BlockLinearOperator<Range, Domain>
+block_operator(const BlockMatrix &block_matrix)
+{
+  typedef LinearOperator<typename Range::BlockType, typename Domain::BlockType> BlockType;
+
+  BlockLinearOperator<Range, Domain> return_op;
+
+  return_op.n_block_rows = [&block_matrix]() -> unsigned int
+  {
+    return block_matrix.n_block_rows();
+  };
+
+  return_op.n_block_cols = [&block_matrix]() -> unsigned int
+  {
+    return block_matrix.n_block_cols();
+  };
+
+  return_op.block = [&block_matrix](unsigned int i, unsigned int j) -> BlockType
+  {
+    const unsigned int m = block_matrix.n_block_rows();
+    const unsigned int n = block_matrix.n_block_cols();
+    Assert(i < m, ExcIndexRange (i, 0, m));
```

Comment:
You'll get warnings about unused `m` and `n` here in optimized mode. Just inline the expressions that currently initialize the variables into the assert.